### PR TITLE
Dissociate volume creation from server attachment

### PR DESCRIPTION
**Pull request by sbesson**
Follow-up of #4

In the end it turns out the upstream configuration changes require the volumes to be flattened at creation which adds an extra couple of hours to the deployment.

To accommodate this new workflow, this PR includes the following breaking changes:
- the attachment logic is now dependent on the presence of `openstack_volume_vmname`
- the snapshot is renamed to drop `openstack_volume_vmname` which is optional and be suffixed by `-source`
- the snapshot is deleted when `openstack_volume_vmname` is set

This allows a typical provisioning workflow to happen in two steps:
- first execute this role with `openstack_volume_source` set but `openstack_volume_vmname` unset. This should trigger the creation of a snapshot and of a derived volume. The second operation will not complete immediately if the volume is flattened
- then re-execute this role with `openstack_volume_vmname`  set. This should clean up the snapshot created in step 1 (and fail if there is still a dependency) and attach the flattened volume to the instance

### CODE REVIEW

**Comment by joshmoore** [comment]
No objections to the logic _per se_ (though one does have to read carefully to understand) but I wonder if this has any potential of surprising someone who is using it. I guess since clean up only occurs if a variable is set, it should be safe.

**Comment by sbesson** [comment]
It would be great to pin down the potential source of confusion and either make changes or clarify the README.

In version 1.2.0 and before, `openstack_volume_vmname` was required and `openstack_volume_source` was optional and the behavior of the role was as follows:

1. create a new volume using a source volume if `openstack_volume_source` is set and wait
2. attach the volume to the instance specified by `openstack_volume_vmname`

With  the current proposal, both `openstack_volume_vmname`  and `openstack_volume_source` are optional with the following outcomes:

- `openstack_volume_vmname` set and `openstack_volume_source` unset: strictly identical behavior as previously
- `openstack_volume_vmname` and `openstack_volume_source` both set: almost identical behavior as previously with the exception that an intermediate snapshot is created and cleaned up as the source of the new volume
- `openstack_volume_source` set and `openstack_volume_vmname` unset: create a source snapshot, create a new volume and return immediately (no wait)
- `openstack_volume_source` unset and `openstack_volume_vmname` unset: create a new volume and return immediately (no wait)

In all cases, `openstack_volume_name` is now the exact name of the volume (breaking change).

**Comment by sbesson** [comment]
After re-reading the README, I am considering introducing a new optional variable `openstack_volume_snapshot_cleanup` to control the snapshot cleanup (and the wait on volume creation) rather than making this logic conditional to the volume attachment.

**Comment by sbesson** [comment]
The last version of the role together with https://github.com/IDR/deployment/pull/358 was successfully used to create `pilot-idr0128`. Hopefully, `openstack_volume_wait` should partly lift the concerns raised above regarding the potential for confusion.

File: README.md [hunk 1]
```diff
@@ -15,19 +16,12 @@ Role Variables
...
-If `openstack_volume_source` is specified, the role will create an intermediate
-volume snapshot, named after `openstack_volume_name` and suffixed with
-`-source`, and use this snapshot as the source of the new volume.
-
-When `openstack_volume_vmname` is passed, the snapshot used for the volume
-creation is cleaned up (if it exists) before attaching the volume to the
-instance.
+- `openstack_volume_device`: The device name of the volume in the instance (default: automatically determined by OpenStack)
+- `openstack_volume_type`: The type of volume (default: automatically determined by OpenStack)
+- `openstack_volume_source`: If defined, the name of an existing volume to use as the source of the volume creation. An temporary volume snapshot will be created, named after `openstack_volume_name` and suffixed with `-source` and cleaned after the volume is created.
```
Comment (joshmoore) [inline README.md]:
```suggestion
- `openstack_volume_source`: If defined, the name of an existing volume to use as the source of the volume creation. A temporary volume snapshot will be created, named after `openstack_volume_name` and suffixed with `-source` and cleaned after the volume is created.
```

Comment (joshmoore) [reply README.md]: It's a bit unclear why the user needs to know that it will be created if it's then deleted.

**Review by joshmoore — COMMENTED** [review]
New changes look good. One comment; one minor typo fix.

File: README.md (hunk 1 above)
Comment (sbesson) [reply README.md]: I agree, this is internal implementation detail and has no user-facing value except in the scenario where `openstack_volume_wait` is False as this role will leave an uncleaned snapshot. [e96b3e2](https://github.com/ome/ansible-role-openstack-volume-storage/pull/5/commits/e96b3e2d4b1b432e82cabf4eb372d94becd3b343) should address this

**Comment by joshmoore** [comment]
Looking better, thanks. Like having all the complexity under `_wait` so someone is careful when/if they use it.

**Pull request merged**